src/ferc1_eia_match/candidate_set_creation.py

```diff
@@ -0,0 +1,344 @@
+        tuple_embedding = tuple_embedding.tocsr()
+        return tuple_embedding
+
+    def set_col_blocks(self, blocking_col: str):
+        """Get the indices of blocks of records in a dataframe for a blocking column.
+
+        Get blocks of records in a dataframe where the value of the ``blocking_col``
+        column is equal within all records in a block. Update ``self.left_blocks_dict`` and
+        ``self.right_blocks_dict`` to be a dictionary with keys as each unique value in
+        the ``blocking_col`` column and values as the index of the rows with that value.
+        This is the result of ``df.groupby(blocking_col).groups``. Update ``self.blocking_col``
+
+        Arguments:
+            df: The dataframe from which to get blocks of records.
+            blocking_col: The name of the column to block on.
+        """
+        self.blocking_col = blocking_col
+        if blocking_col in self.left_df.columns:
+            self.left_blocks_dict = self.left_df.groupby(blocking_col).groups
+        if blocking_col in self.right_df.columns:
+            self.right_blocks_dict = self.right_df.groupby(blocking_col).groups
+
+    def embed_dataframes(self, blocking_col: str = "") -> None:
+        """Embed left (and right) dataframes and create blocks from a column.
+
+        Set `self.left_embedding_matrix` and `self.right_embedding_matrix` with
+        matrix embeddings for `self.left_df` and `self.right_df`. Embed attributes
+        based on the functions for each column in `self.col_embedding_dict`.
+        Concatenate the embeddings for each column together into one embedding
+        matrix for each dataframe. Optionally set `self.left_blocks_dict` and
+        ``self.right_blocks_dict`` to be the indices of blocks within ``blocking_col``.
+
+        Arguments:
+            blocking_col: The name of the column to get blocks of indices of the
+                dataframes. Default is the empty string if no manual blocking on
+                the value of a column is to be performed. Passing a blocking column
+                will not modify the dataframes themselves, just set the respective
+                dictionary of blocks.
+        """
+        # TODO: there's probably a better way to do this
+        for column_name, vectorizer in self.col_embedding_dict.items():
+            if vectorizer == "tfidf":
+                self.tfidf_vectorize(column_name=column_name)
+            elif vectorizer == "min_max_scale":
+                self.min_max_scale(column_name=column_name)
+        # TODO: allow for other types of aggregation
+        self.left_embedding_matrix = self.equal_weight_aggregate(
+            self.left_embedding_attribute_dict
+        )
+        self.right_embedding_matrix = self.equal_weight_aggregate(
+            self.right_embedding_attribute_dict
+        )
+        if blocking_col != "":
+            self.set_col_blocks(blocking_col=blocking_col)
+
+
+class SimilaritySearcher:
```

Comment:
This class conducts the similarity search, in this case finding k best EIA records for each FERC record. 

The query embedding matrix is the records that you want to find the k best of the other side for. In our case, FERC.
The menu embedding matrix is the records from which to search for the k best from. In our case, EIA.

I implemented raw L2 distance minimization search as well as maximizing cosine similarity search (same as taking the inner product of normalized vectors).

Reply:
In the original comment above you mentioned that there was 95% recall when the query vector was FERC but 100% recall when it was EIA, but we need for it to be FERC right? Since those are the "real" records, and the EIA side of things is all synthetic records?